query/physicalplan/synchronize.go

```diff
@@ -0,0 +1,55 @@
+package physicalplan
+
+import (
+	"context"
+	"sync"
+
+	"github.com/apache/arrow/go/v8/arrow"
+	"go.uber.org/atomic"
+)
+
+// Synchronizer is used to combine the results of multiple parallel streams
+// into a single stream concurrent stream. It also forms a barrier on the
+// finishers, by waiting to call next plan's finish until all previous parallel
+// stages have finished.
+type Synchronizer struct {
+	next    PhysicalPlan
+	nextMtx sync.Mutex
+	running *atomic.Uint64
+}
+
+func Synchronize(concurrency int) *Synchronizer {
+	return &Synchronizer{running: atomic.NewUint64(uint64(concurrency))}
+}
+
+func (m *Synchronizer) Callback(ctx context.Context, r arrow.Record) error {
+	// multiple threads can emit the results to the next step, but they will do
+	// it synchronously
+	m.nextMtx.Lock()
```

Comment:
Instead of using a mutex, I would use a buffered channel here (maybe with `concurrency` slots) with a synchronizer goroutine reading from the channel and pushing. The nice thing with a buffered channel is that upstream goroutines can push data to be read even if the downstream goroutine is not ready to read this data yet (e.g. so the upstream goroutine can perform more scan work concurrently), while still being able to apply backpressure if the downstream goroutine is super slow for some reason.

Reply:
Yes, I had that in between implementing but reverted it since there were too many moving things at once. That 💯 a follow up for this PR.